pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,1015 @@
+				}
+			}
+		}
+		if ipTypeMap.Has(v1alpha1.ServiceIPTypeLoadBalancerIP) && svc.Spec.Type == corev1.ServiceTypeLoadBalancer {
+			if externalLocal && hasLocalEndpoints || !externalLocal {
+				for _, ingressIP := range svc.Status.LoadBalancer.Ingress {
+					if ingressIP.IP != "" {
+						serviceIPs = append(serviceIPs, ingressIP.IP)
+					}
+				}
+			}
+		}
+	}
+
+	for _, ip := range serviceIPs {
+		if c.enabledIPv4 && net.IsIPv4String(ip) {
+			allRoutes.Insert(bgp.Route{Prefix: ip + ipv4Suffix})
+		}
+		if c.enabledIPv6 && net.IsIPv6String(ip) {
+			allRoutes.Insert(bgp.Route{Prefix: ip + ipv6Suffix})
+		}
+	}
+
+	return nil
+}
+
+func (c *Controller) addEgressRoutes(allRoutes sets.Set[bgp.Route]) error {
+	egresses, err := c.egressLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	for _, eg := range egresses {
+		if eg.Status.EgressNode != c.nodeName {
+			continue
+		}
+		ip := eg.Status.EgressIP
+		if c.enabledIPv4 && net.IsIPv4String(ip) {
+			allRoutes.Insert(bgp.Route{Prefix: ip + ipv4Suffix})
+		}
+		if c.enabledIPv6 && net.IsIPv6String(ip) {
+			allRoutes.Insert(bgp.Route{Prefix: ip + ipv6Suffix})
+		}
+	}
+
+	return nil
+}
+
+func (c *Controller) addPodRoutes(allRoutes sets.Set[bgp.Route]) {
+	if c.enabledIPv4 {
+		allRoutes.Insert(bgp.Route{Prefix: c.podIPv4CIDR})
+	}
+	if c.enabledIPv6 {
+		allRoutes.Insert(bgp.Route{Prefix: c.podIPv6CIDR})
+	}
+}
+
+func (c *Controller) hasLocalEndpoints(svc *corev1.Service) (bool, error) {
+	eps, err := c.endpointSliceLister.EndpointSlices(svc.GetNamespace()).Get(svc.GetName())
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+
+	for _, ep := range eps.Endpoints {
+		if ep.NodeName != nil && *ep.NodeName == c.nodeName {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func (c *Controller) generateBGPPeerConfig(peer *v1alpha1.BGPPeer) bgp.PeerConfig {
+	bgpPeerConfig := bgp.PeerConfig{
+		BGPPeer: peer,
+	}
+	bgpPeerKey := generateBGPPeerKey(peer.Address, peer.ASN)
+	c.bgpPeerPasswordsMutex.RLock()
+	defer c.bgpPeerPasswordsMutex.RUnlock()
+	if password, exists := c.bgpPeerPasswords[bgpPeerKey]; exists {
+		bgpPeerConfig.Password = password
+	}
+	return bgpPeerConfig
+}
+
+func (c *Controller) getPeers(allPeers []v1alpha1.BGPPeer) (sets.Set[string], map[string]bgp.PeerConfig, error) {
+	peerKeys := sets.New[string]()
+	peerConfigs := make(map[string]bgp.PeerConfig)
+
+	for i := range allPeers {
+		peerKey := generateBGPPeerKey(allPeers[i].Address, allPeers[i].ASN)
+		if c.enabledIPv4 && net.IsIPv4String(allPeers[i].Address) {
+			peerKeys.Insert(peerKey)
+		}
+		if c.enabledIPv6 && net.IsIPv6String(allPeers[i].Address) {
+			peerKeys.Insert(peerKey)
+		}
+		peerConfigs[peerKey] = c.generateBGPPeerConfig(&allPeers[i])
+	}
+	return peerKeys, peerConfigs, nil
+}
+
+func generateBGPPeerKey(address string, asn int32) string {
+	return fmt.Sprintf("%s-%d", address, asn)
```

Comment:
Can this work well with IPv6? Can the secret key contain string like '::'?

Reply:
Yes.